File: android/app/src/main/java/com/noop/push/ServerVitalSelection.kt
package com.noop.push

/** Once a metric is owned, null, pending and failure remain server results. */
data class ServerVitalSelection(
    val value: Double?, val fromServer: Boolean, val day: String, val status: String?, val stale: Boolean,
    val sourceFeature: String? = null, val deviceId: String? = null, val algorithmVersion: String? = null,
    val resultRevision: String? = null,
) {
    val displayDiagnostic: ServerScoreStageDiagnostic get() = ServerScoreStageDiagnostic("displayed",
        if (value != null) "available" else "unavailable", if (!fromServer) "local_producer_retained"
            else if (value == null) "metric_unavailable" else "server_metric_selected")
    enum class Metric(val key: String, val feature: String) {
        HRV("hrv_rmssd_ms", "hrv"), RESTING_HR("resting_hr_bpm", "hrv"),
        RESPIRATORY("resp_rate_bpm", "respiration"), SLEEP("sleep_total_min", "sleep"), REST("sleep_performance", "sleep"),
        CHARGE("recovery", "hrv"), STRAIN("strain", "hrv"), SPO2("spo2_pct", "hrv"),
        SKIN_TEMP("skin_temp_c", "hrv")
    }

    companion object {
        /** [overlay] must already be owner scoped and configuration/authentication qualified. */
        fun resolve(metric: Metric, serverEnabled: Boolean, selectedDay: String,
                    overlay: ServerScoreDayCache?, localValue: Double?): ServerVitalSelection {
            val family = overlay?.takeIf { it.day == selectedDay }?.compute?.familyFor(metric.key)
            if (family != null) return ServerVitalSelection(family.number(metric.key), true, selectedDay,
                overlay.readFailure ?: family.reason ?: family.status, overlay.stale || family.status == "stale",
                family.family, family.deviceId, family.algorithmVersion, family.resultRevision)
            val feature = overlay?.features?.get(metric.feature)
            val owned = overlay?.ownedMetrics?.contains(metric.key) ?: serverEnabled
            if (!owned) return ServerVitalSelection(localValue, false, selectedDay, null, false)
            if (overlay == null || overlay.day != selectedDay)
                return ServerVitalSelection(null, true, selectedDay, "pending", false, metric.feature)
            val (value, featureKey) = when (metric) {
                Metric.HRV -> overlay.daily?.hrvRmssdMs to "hrv"
                Metric.RESTING_HR -> overlay.daily?.restingHrBpm?.toDouble() to "hrv"
                Metric.RESPIRATORY -> overlay.daily?.respRateBpm to "respiration"
                Metric.SLEEP -> overlay.daily?.sleepTotalMin to "sleep"
                Metric.REST -> overlay.daily?.rest to "sleep"
                Metric.CHARGE -> overlay.daily?.recovery to "hrv"
                Metric.STRAIN -> overlay.daily?.strain to "hrv"
                Metric.SPO2 -> overlay.daily?.spo2Pct to "hrv"
                Metric.SKIN_TEMP -> overlay.daily?.skinTempC to "hrv"
            }
            val status = feature?.status ?: "unavailable"
            val available = feature?.isCanonicalAvailable == true
            if (!available || value == null) {
                val missingState = overlay.readFailure ?: feature?.processingStatus?.takeIf { it in setOf("pending", "running", "retry", "failed", "exhausted") }
                    ?: feature?.reason ?: if (value == null && available) "insufficient_input" else "unavailable"
                return ServerVitalSelection(null, true, selectedDay, missingState, overlay.stale,
                    feature?.let { featureKey }, feature?.deviceId, feature?.algorithmVersion)
            }
            return ServerVitalSelection(value, true, selectedDay, status, overlay.stale || status == "stale",
                feature?.let { featureKey }, feature?.deviceId, feature?.algorithmVersion)
        }
    }
}

File: Packages/WhoopStore/Sources/WhoopStore/ServerVitalSelection.swift
import Foundation
import WhoopProtocol

/// Hosted mode retains explicit server ownership for computed physiology metrics.
/// The caller supplies an owner-scoped overlay only after configuration and authentication checks.
public struct ServerVitalSelection: Equatable {
    public enum Metric: CaseIterable {
        case hrv, restingHR, respiratory, sleep, rest, charge, strain, spo2, skinTemp
        var key: String {
            switch self {
            case .hrv: return "hrv_rmssd_ms"
            case .restingHR: return "resting_hr_bpm"
            case .respiratory: return "resp_rate_bpm"
            case .sleep: return "sleep_total_min"
            case .rest: return "sleep_performance"
            case .charge: return "recovery"
            case .strain: return "strain"
            case .spo2: return "spo2_pct"
            case .skinTemp: return "skin_temp_c"
            }
        }
        var feature: String {
            switch self {
            case .respiratory: return "respiration"
            case .sleep, .rest: return "sleep"
            default: return "hrv"
            }
        }
    }
    public let value: Double?
    public let fromServer: Bool
    public let day: String
    public let status: String?
    public let stale: Bool
    public let sourceFeature: String?
    public let deviceId: String?
    public let algorithmVersion: String?
    public var canonicalResult: ServerCanonicalFamilyResult?
    public var displayDiagnostic: ServerScoreStageDiagnostic {
        ServerScoreStageDiagnostic(stage: "displayed", status: value == nil ? "unavailable" : "available",
            reason: !fromServer ? "local_producer_retained" : value == nil ? "metric_unavailable" : "server_metric_selected")
    }

    public static func resolve(_ metric: Metric, serverEnabled: Bool, selectedDay: String,
                               overlay: ServerScoreDayCache?, localValue: @autoclosure () -> Double?) -> Self {
        if PhoneComputeRuntime.isFinalHosted || overlay?.canonicalResults != nil || overlay?.pendingCanonicalResults != nil {
            let family = overlay?.day == selectedDay ? overlay?.canonicalResults?.result(for: metric.key) : nil
            let pending = overlay?.day == selectedDay ? overlay?.pendingCanonicalResults : nil
            var result = Self(value: family?.number(metric.key), fromServer: true, day: selectedDay,
                status: overlay?.readFailure ?? family?.missingReason(metric.key) ?? family?.reason ?? family?.status ?? pending?.reason ?? "awaiting_server_result",
                stale: overlay?.stale == true || family?.freshness != "current",
                sourceFeature: metric.feature, deviceId: family?.deviceID, algorithmVersion: family?.algorithmVersion)
            result.canonicalResult = family
            return result
        }
        // An explicit ledger supersedes the legacy caller's hosted-selection argument.
        let owned = overlay?.ownedMetrics?.contains(metric.key) ?? serverEnabled
        guard owned else {
            return Self(value: localValue(), fromServer: false, day: selectedDay, status: nil, stale: false,
                        sourceFeature: nil, deviceId: nil, algorithmVersion: nil)
        }
        // A pending snapshot must not resurrect a competing local calculation.
        guard let overlay, overlay.day == selectedDay else {
            return Self(value: nil, fromServer: true, day: selectedDay, status: "pending", stale: false,
                        sourceFeature: metric.feature, deviceId: nil, algorithmVersion: nil)
        }
        let value: Double?
        let featureKey: String
        switch metric {
        case .hrv: value = overlay.daily?.hrvRmssdMs; featureKey = "hrv"
        case .restingHR: value = overlay.daily?.restingHrBpm.map(Double.init); featureKey = "hrv"
        case .respiratory: value = overlay.daily?.respRateBpm; featureKey = "respiration"
        case .sleep: value = overlay.daily?.sleepTotalMin; featureKey = "sleep"
        case .rest: value = overlay.daily?.rest; featureKey = "sleep"
        case .charge: value = overlay.daily?.recovery; featureKey = "hrv"
        case .strain: value = overlay.daily?.strain; featureKey = "hrv"
        case .spo2: value = overlay.daily?.spo2Pct; featureKey = "hrv"
        case .skinTemp: value = overlay.daily?.skinTempC; featureKey = "hrv"
        }
        let feature = overlay.features[featureKey]
        let status = feature?.status ?? "unavailable"
        let available = feature?.isCanonicalAvailable == true
        // Missing/unqualified server results remain unavailable rather than falling back locally.
        if !available || value == nil {
            let legacyReason = feature?.algorithmVersion == ServerLegacyReadEligibility.algorithm &&
                (ServerLegacyReadEligibility.always.contains(metric.key) ||
                 overlay.legacySleepWithheld && (ServerLegacyReadEligibility.sleep.contains(metric.key) || metric == .rest))
                ? ServerLegacyReadEligibility.reason : nil
            let processing = feature?.processingStatus.flatMap {
                ["pending", "running", "retry", "failed", "exhausted"].contains($0) ? $0 : nil
            }
            return Self(value: nil, fromServer: true,
                        day: selectedDay, status: overlay.readFailure ?? legacyReason ?? processing ?? feature?.reason ?? (available ? "insufficient_input" : "unavailable"), stale: overlay.stale,
                        sourceFeature: feature == nil ? nil : featureKey, deviceId: feature?.deviceId,
                        algorithmVersion: feature?.algorithmVersion)
        }
        return Self(value: value, fromServer: true, day: selectedDay,
                    status: status,
                    stale: overlay.stale || status == "stale", sourceFeature: feature == nil ? nil : featureKey,
                    deviceId: feature?.deviceId, algorithmVersion: feature?.algorithmVersion)
    }
}
